targets/esp8266/jshardware.c

```diff
@@ -692,8 +666,19 @@ bool jshIsUSBSERIALConnected() {
  */
 void jshUSARTKick(
     IOEventFlags device //!< The device to be kicked.
-  ) {
-  esp8266_uartTransmitAll(device);
+) {
+  // transmit anything that is sitting in the uart output buffers
+  if (device == EV_SERIAL1 || device == EV_SERIAL2) {
+    // Get the next character to transmit.  We will have reached the end when
+    // the value of the character to transmit is -1.
+    int c = jshGetCharToTransmit(device);
```

Comment:
There does actually appear to be some UART TX empty interrupt [like this](https://github.com/espressif/ESP8266_RTOS_SDK/blob/master/examples/driver_lib/driver/uart.c#L390) but I don't know if it's exposed at all? If so, that might really help here?